packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -327,6 +374,55 @@ export default function makeKernelKeeper(kernelStorage, kernelSlog) {
     initializeStats();
   }
 
+  // maybeUpgradeKernelState() can be called multiple times; this flag
+  // helps avoid extra "are we upgraded" DB queries
+  let upgraded = false;
+
+  // the caller is responsible for ensuring that any changes made here
+  // get committed at a useful point
```

Comment:
Yeah, there's one `upgraded` flag per `KernelKeeper` object, and each kernel has exactly one `KernelKeeper` (created at the start of `buildKernel()`), but we also create a `KernelKeeper` when we initialize/populate the DB, and we should *not* be doing an upgrade at that time, which is why `maybeUpgradeKernelState()` is a method and not an automatic consequence of `makeKernelKeeper()`.

It's currently called just before each call to `startCrank()`.. I wanted to make it more explicit, because (as you noted) it's important to call it from outside a savepoint. I was worried that adding it to `startCrank()` would obscure the upgrade point and make it fragile under maintenance.

But.. in looking more carefully at how/when `startCrank` is called (it also show up in `changeKernelOptions` and `installBundle`, to which I did *not* add a `maybeUpgrade` call), I think my PR would have a small bug. `kernelKeeper.startCrank()` calls `kernelStorage.startCrank()` in swing-store, which sets an `inCrank = true` flag, and calls `resetCrankHash()`. By not calling `startCrank` *before* `maybeUpgrade` (or not calling it at all, if we start with `changeKernelOptions` or `installBundle`, which sounds pretty plausible TBH), we don't include any upgrade-related kvStore changes in the crankhash. That's not the intent: upgrades are in-consensus, all nodes of a consensus kernel should upgrade at the same time, and their changes should be included in the crankhash.

So I think I need to change this as you suggest: call `maybeUpgradeKernelState()` from inside `kernelKeeper.startCrank()`, but do it specifically *after* that function calls `kernelStorage.startCrank()`. I'm not excited to conceal the fact that upgrades are happening from call sites like `step()` and `run()`, but it's more important to have these upgrade changes be captured in the hash reliably.

Reply:
> we also create a `KernelKeeper` when we initialize/populate the DB, and we should _not_ be doing an upgrade at that time, which is why `maybeUpgradeKernelState()` is a method and not an automatic consequence of `makeKernelKeeper()`.

That had totally skipped my mind. Can we document somewhere these constraints?